src/py/flwr/client/grpc_rere_client/connection.py

```diff
@@ -107,47 +115,83 @@ def grpc_request_response(
         max_message_length=max_message_length,
     )
     channel.subscribe(on_channel_state_change)
-    stub = FleetStub(channel)
-
-    # Necessary state to validate messages to be sent
-    state: Dict[str, Optional[Metadata]] = {KEY_METADATA: None}
 
-    # Enable create_node and delete_node to store node
-    node_store: Dict[str, Optional[Node]] = {KEY_NODE: None}
+    # Shared variables for inner functions
+    stub = FleetStub(channel)
+    metadata: Optional[Metadata] = None
+    node: Optional[Node] = None
+    ping_thread: Optional[threading.Thread] = None
+    ping_stop_event = threading.Event()
 
     ###########################################################################
-    # receive/send functions
+    # ping/create_node/delete_node/receive/send functions
     ###########################################################################
 
+    def ping() -> None:
+        # Get Node
+        if node is None:
+            log(ERROR, "Node instance missing")
+            if not ping_stop_event.is_set():
+                ping_stop_event.wait(PING_CALL_TIMEOUT)
+            return
+
+        # Construct the ping request
+        req = PingRequest(node=node, ping_interval=PING_DEFAULT_INTERVAL)
+
+        # Call FleetAPI
+        res: PingResponse = stub.Ping(req, timeout=PING_CALL_TIMEOUT)
```

Comment:
If an `RpcError` is raised here, we're not waiting until the `next_interval`. This is desired for cases where the ping call fails because of a network failure. But if, for example, the server is down, then an immediate retry would lead to a number of client nodes retrying in a busy loop. Once the server restarts, it could be brought down immediately because all nodes will retry their ping requests at the same time. To prevent that, we should probably have one immediate retry followed by retries on an exponential backoff schedule.